pkg/ostree/app.js

```diff
@@ -200,30 +236,126 @@
         }])
 
         .directive('ostreeCheck', [
-            function() {
+            '$modal',
+            "remoteActions",
+            function($modal, remoteActions) {
                 return {
                     restrict: 'E',
                     templateUrl: "ostree-check.html",
                     scope: {
                         os: "=",
-                        runningMethod: "="
+                        runningMethod: "=",
+                        currentOrigin: "="
                     },
                     link: function(scope, element, attrs) {
                         scope.error = null;
                         scope.progressMsg = null;
                         scope.isRunning = false;
+                        scope.branches = {};
+
+                        function updateBranchCache(remote, branch) {
+                            scope.isRunning = true;
+
+                            /* We want to populate the list with already downloaded updates
+                             * but this can error if there are no new updates or the remote
+                             * hasn't yes been downloaded from so we ignore errors.
+                             * If there is a real error, the use will see if when they
+                             * check for updates.
+                             */
+                            client.cache_update_for(scope.os,
+                                                    remote,
+                                                    branch)
+                                .always(function () {
+                                    scope.$applyAsync(function () {
+                                        scope.isRunning = false;
+                                    });
+                                });
+                        }
+
+                        scope.$watch("currentOrigin.remote", function (newValue) {
+                            var branches = {
+                                remote: newValue
+                            };
+
+                            if (!newValue) {
+                                scope.branches = {};
+                                return;
+                            }
+
+                            // Only run when remote is changed by loading new OS
+                            // dialog takes care of this when changed there
+                            if (scope.branches.remote === newValue)
+                                return;
+
+                            scope.isRunning = true;
+                            remoteActions.listBranches(newValue)
+                                .then(function (data) {
+                                    branches.list = data;
+                                }, function (ex) {
+                                    branches.error = cockpit.message(ex);
```

Comment:
Since branches is created as a new object and only assigned to the scope after the promise is complete. ```branches.list``` will always be undefined in the error case.